File: proxy_logger/logger.py
import json
import logging
import os
import threading
import time
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from logging.handlers import RotatingFileHandler
from pathlib import Path
from typing import Optional

logger = logging.getLogger("proxy_logger")


@dataclass
class LogEntry:
    id: str = ""
    timestamp: str = ""
    duration_ms: float = 0
    client_addr: str = ""

    # Request
    method: str = ""
    url: str = ""
    host: str = ""
    request_headers: dict = field(default_factory=dict)
    request_body: str = ""
    request_size: int = 0

    # Response
    status_code: int = 0
    response_headers: dict = field(default_factory=dict)
    response_body: str = ""
    response_size: int = 0

    # TLS
    tls: bool = False
    tls_version: str = ""

    # Error
    error: str = ""

    def to_json(self) -> str:
        d = asdict(self)
        # Remove empty optional fields
        return json.dumps(
            {k: v for k, v in d.items() if v or k in ("tls", "status_code", "request_size", "response_size")},
            ensure_ascii=False,
        )

    def to_text(self) -> str:
        status = str(self.status_code) if self.status_code else "ERR"
        return (
            f"[{self.timestamp}] {self.client_addr} {self.method} {self.url} "
            f"-> {status} ({self.duration_ms:.0f}ms, req:{self.request_size} resp:{self.response_size})"
        )


class Stats:
    def __init__(self):
        self._lock = threading.Lock()
        self.total_requests: int = 0
        self.total_bytes: int = 0
        self.active_conns: int = 0
        self.error_count: int = 0
        self.start_time: float = time.time()
        self.recent_entries: list[dict] = []
        self._max_recent = 200

    def record(self, entry: LogEntry):
        with self._lock:
            self.total_requests += 1
            self.total_bytes += entry.request_size + entry.response_size
            if entry.error:
                self.error_count += 1
            self.recent_entries.append(json.loads(entry.to_json()))
            if len(self.recent_entries) > self._max_recent:
                self.recent_entries = self.recent_entries[-self._max_recent:]

    def conn_open(self):
        with self._lock:
            self.active_conns += 1

    def conn_close(self):
        with self._lock:
            self.active_conns -= 1

    def to_dict(self) -> dict:
        with self._lock:
            return {
                "total_requests": self.total_requests,
                "total_bytes": self.total_bytes,
                "active_conns": self.active_conns,
                "error_count": self.error_count,
                "uptime_seconds": int(time.time() - self.start_time),
            }


class ProxyLogger:
    """Structured logger with file rotation for proxy traffic."""

    def __init__(
        self,
        log_dir: str = "logs",
        max_size_mb: int = 100,
        max_backups: int = 10,
        log_format: str = "json",
    ):
        self.log_format = log_format
        self.stats = Stats()

        # Setup log directory
        Path(log_dir).mkdir(parents=True, exist_ok=True)

        # Setup rotating file handler
        log_file = os.path.join(log_dir, "proxy.log")
        self._handler = RotatingFileHandler(
            log_file,
            maxBytes=max_size_mb * 1024 * 1024,
            backupCount=max_backups,
            encoding="utf-8",
        )
        self._handler.setFormatter(logging.Formatter("%(message)s"))

        self._logger = logging.getLogger("proxy_traffic")
        self._logger.setLevel(logging.INFO)
        self._logger.addHandler(self._handler)
        # Prevent propagation to root logger
        self._logger.propagate = False

    def log(self, entry: LogEntry):
        """Log a proxy traffic entry."""
        self.stats.record(entry)

        if self.log_format == "text":
            self._logger.info(entry.to_text())
        else:
            self._logger.info(entry.to_json())

    def close(self):
        self._handler.close()

File: internal/logger/logger.go
package logger

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"sync/atomic"
	"time"
)

// Entry represents a single proxied HTTP transaction log entry.
type Entry struct {
	ID        string    `json:"id"`
	Timestamp time.Time `json:"timestamp"`
	Duration  int64     `json:"duration_ms"`
	Client    string    `json:"client_addr"`

	// Request
	Method     string            `json:"method"`
	URL        string            `json:"url"`
	Host       string            `json:"host"`
	ReqHeaders map[string]string `json:"request_headers"`
	ReqBody    string            `json:"request_body,omitempty"`
	ReqSize    int64             `json:"request_size"`

	// Response
	StatusCode  int               `json:"status_code"`
	RespHeaders map[string]string `json:"response_headers,omitempty"`
	RespBody    string            `json:"response_body,omitempty"`
	RespSize    int64             `json:"response_size"`

	// TLS info
	TLS        bool   `json:"tls"`
	TLSVersion string `json:"tls_version,omitempty"`

	// Error
	Error string `json:"error,omitempty"`
}

// Stats tracks proxy statistics.
type Stats struct {
	TotalRequests   atomic.Int64
	TotalBytes      atomic.Int64
	ActiveConns     atomic.Int64
	ErrorCount      atomic.Int64
	StartTime       time.Time
}

// Logger handles structured logging with file rotation.
type Logger struct {
	mu          sync.Mutex
	dir         string
	maxSizeMB   int
	maxBackups  int
	format      string
	currentFile *os.File
	currentSize int64

	Stats Stats
}

// New creates a new Logger.
func New(dir string, maxSizeMB, maxBackups int, format string) (*Logger, error) {
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return nil, fmt.Errorf("create log dir: %w", err)
	}

	l := &Logger{
		dir:        dir,
		maxSizeMB:  maxSizeMB,
		maxBackups: maxBackups,
		format:     format,
	}
	l.Stats.StartTime = time.Now()

	if err := l.openFile(); err != nil {
		return nil, err
	}

	return l, nil
}

func (l *Logger) openFile() error {
	filename := filepath.Join(l.dir, "proxy.log")
	f, err := os.OpenFile(filename, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return fmt.Errorf("open log file: %w", err)
	}

	info, err := f.Stat()
	if err != nil {
		f.Close()
		return err
	}

	l.currentFile = f
	l.currentSize = info.Size()
	return nil
}

func (l *Logger) rotate() error {
	if l.currentFile != nil {
		l.currentFile.Close()
	}

	// Rotate existing backups
	base := filepath.Join(l.dir, "proxy.log")
	for i := l.maxBackups - 1; i >= 1; i-- {
		src := fmt.Sprintf("%s.%d", base, i)
		dst := fmt.Sprintf("%s.%d", base, i+1)
		os.Rename(src, dst)
	}
	os.Rename(base, base+".1")

	// Remove oldest backup if exceeds maxBackups
	oldest := fmt.Sprintf("%s.%d", base, l.maxBackups+1)
	os.Remove(oldest)

	return l.openFile()
}

// Log writes an entry to the log file.
func (l *Logger) Log(entry *Entry) error {
	l.Stats.TotalRequests.Add(1)
	l.Stats.TotalBytes.Add(entry.ReqSize + entry.RespSize)
	if entry.Error != "" {
		l.Stats.ErrorCount.Add(1)
	}

	var line []byte
	var err error

	switch l.format {
	case "text":
		line = []byte(formatText(entry))
	default:
		line, err = json.Marshal(entry)
		if err != nil {
			return fmt.Errorf("marshal log entry: %w", err)
		}
		line = append(line, '\n')
	}

	l.mu.Lock()
	defer l.mu.Unlock()

	maxBytes := int64(l.maxSizeMB) * 1024 * 1024
	if l.currentSize+int64(len(line)) > maxBytes {
		if err := l.rotate(); err != nil {
			return fmt.Errorf("rotate log: %w", err)
		}
	}

	n, err := l.currentFile.Write(line)
	if err != nil {
		return fmt.Errorf("write log: %w", err)
	}
	l.currentSize += int64(n)

	return nil
}

func formatText(e *Entry) string {
	status := fmt.Sprintf("%d", e.StatusCode)
	if e.Error != "" {
		status = "ERR"
	}
	return fmt.Sprintf("[%s] %s %s %s -> %s (%dms, req:%d resp:%d)\n",
		e.Timestamp.Format("2006-01-02 15:04:05.000"),
		e.Client,
		e.Method,
		e.URL,
		status,
		e.Duration,
		e.ReqSize,
		e.RespSize,
	)
}

// Close closes the logger.
func (l *Logger) Close() error {
	l.mu.Lock()
	defer l.mu.Unlock()
	if l.currentFile != nil {
		return l.currentFile.Close()
	}
	return nil
}

// GetStats returns current statistics as a map.
func (l *Logger) GetStats() map[string]interface{} {
	return map[string]interface{}{
		"total_requests": l.Stats.TotalRequests.Load(),
		"total_bytes":    l.Stats.TotalBytes.Load(),
		"active_conns":   l.Stats.ActiveConns.Load(),
		"error_count":    l.Stats.ErrorCount.Load(),
		"uptime_seconds": int64(time.Since(l.Stats.StartTime).Seconds()),
	}
}
